odbc/src/api/diag.rs

```diff
@@ -0,0 +1,159 @@
+use crate::{
+    api::{data::i16_len, definitions::DiagType},
+    errors::ODBCError,
+};
+use odbc_sys::Pointer;
+use odbc_sys::{Char, Integer, SmallInt, SqlReturn, WChar};
+use std::ptr::copy_nonoverlapping;
+
+///
+/// set_sql_state writes the given sql state to the [`output_ptr`].
+///
+/// # Safety
+/// This writes to a raw C-pointer
+///
+pub unsafe fn set_sql_state(sql_state: &str, output_ptr: *mut WChar) {
+    if output_ptr.is_null() {
+        return;
+    }
+    let sql_state = &format!("{}\0", sql_state);
+    let state_u16 = sql_state.encode_utf16().collect::<Vec<u16>>();
+    copy_nonoverlapping(state_u16.as_ptr(), output_ptr, 6);
+}
+
+///
+/// get_diag_rec copies the given ODBC error's diagnostic information
+/// into the provided pointers.
+///
+/// # Safety
+/// This writes to multiple raw C-pointers
+///
+pub unsafe fn get_diag_rec(
+    error: &ODBCError,
+    state: *mut WChar,
+    message_text: Pointer,
+    buffer_length: SmallInt,
+    text_length_ptr: *mut SmallInt,
+    native_error_ptr: *mut Integer,
+    wstring: bool,
+) -> SqlReturn {
+    if !native_error_ptr.is_null() {
+        *native_error_ptr = error.get_native_err_code();
+    }
+    set_sql_state(error.get_sql_state(), state);
```

Comment:
Here, you'll need to call the right set_sql_state, depending of the value for wstring because sqlstate is a string.